---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# micromeg

<!-- badges: start -->
<!-- badges: end -->

The goal of `micromeg` is to document and share convenience functions I use frequently for microbiome data analysis and processing. This package does not create any brand new functionatlity and was built on the great work of others. Many of the functions it provides already exist in other packages. For example, [phyloseq](https://bioconductor.org/packages/release/bioc/html/phyloseq.html) provides many similar tools, and is very well-documented and commonly used (I use it myself!) and so may be better for your purposes.    
    
Other R packages heavily used here include:    
1. [tidyverse](https://tidyverse.tidyverse.org) ecosystem   
2. [vegan](https://cran.r-project.org/web/packages/vegan/index.html)   
3. [dada2](https://benjjneb.github.io/dada2/)   
4. [maaslin3](https://huttenhower.sph.harvard.edu/maaslin3/)        


I started my career in microbiome research at the bench and had to [ELI5](https://www.dictionary.com/e/slang/eli5/) to myself how to process and analyze "big data". I've spent a ton of time poring over and experimenting with [others' code](https://github.com/extrasmallwinnie/micromeg/?tab=readme-ov-file#acknowledgements). Likely, the ideal candidate to benefit from `micromeg` would be another bench scientist without much formal statistics or bioinformatics training. Fair warning, if you already have a strong stats/informatics background, this may not be of much use for you!    

## Installation

You can install the development version of micromeg from [GitHub](https://github.com/extrasmallwinnie/micromeg/) with:

``` r
# install.packages("pak")
pak::pak("extrasmallwinnie/micromeg")
```

## My Usual Workflow

I've listed the steps that I usually take with 16S data, some of which will be demonstrated below with some made up data.    

1. Do the lab work to extract DNA, make libraries, submit for sequencing, etc. Both negative and positive controls are very important! Why, and what are those? Read more here: ^[1](https://pubmed.ncbi.nlm.nih.gov/25387460/),^ ^[2](https://pubmed.ncbi.nlm.nih.gov/27239228/),^ ^[3](https://bmcmicrobiol.biomedcentral.com/articles/10.1186/s12866-020-01839-y),^ ^[4](https://bmcmicrobiol.biomedcentral.com/articles/10.1186/s12866-016-0738-z),^ ^[5](https://journals.asm.org/doi/10.1128/msystems.00062-16)^        
2. Get the data from the sequencing core.   
3. Process the data through [dada2](https://benjjneb.github.io/dada2/) then [decomtam](https://benjjneb.github.io/decontam/vignettes/decontam_intro.html) to generate a cleaned [ASV (amplicon sequencing variant) count table and its associated taxonomy table](https://www.nature.com/articles/nmeth.3869).   
4. Create a "metadata" file with pertinent information on the samples and controls in your run. [Jump to example &darr;](https://github.com/extrasmallwinnie/micromeg/?tab=readme-ov-file#metadata)
5. Do some basic sanity checking on the metadata, ASV, and taxonomy objects. [Jump to example &darr;](https://github.com/extrasmallwinnie/micromeg/?tab=readme-ov-file#sanity-check-the-data)    
6. Check the quality of the sequencing data by examining both the positive and negative controls, and additionally how the controls compare to your real samples.      
7. Alpha diversity (vegan).   
8. Beta diversity (vegan).   
9. Differential abundance (maaslin3).   


```{r empty}
```
## Toy Example

First, load in a very simple example to get an idea for the format of what's expected and the general processing flow. 
```{r empty}
```

### Load in example data
     
I've made up an example study where nasal swabs were taken from people who were either "healthy" or "sick" at the time of sampling. We've collected from each participant their health status, their age at collection, sex, and a simplified location category (in this case, 'rural' or 'urban'). There was also one lab negative control and one lab positive control. 

#### Metadata

Let's load the example metadata object into our session as an object called "metadata":

```{r examplemeta}
library(micromeg)

metadata <- makeExampleMeta()

metadata

```

This metadata object is a tibble (because I personally prefer the [tidyverse grammar](https://www.tmwr.org/tidyverse)), but data frames should also be fine to use. If you use my tools, everything will end up getting \~ tibblefied \~ anyway.     
    
The "SampleID" field wasn't discussed yet, but it's exactly what it sounds like! It must be unique, and it must match what you've called your samples in your sequencing data. (If you use this package, it's highly recommended that you call your sample IDs field exactly "SampleID" in both your metadata and ASV count table objects.)

You may also notice that there's some missing data (the NAs), which we'll talk about more later.

#### ASV count and taxonomy tables
    
In this made up toy example, we did 16S sequencing targeting the V4 region (following [the Kozich et al. 2013 protocol](https://journals.asm.org/doi/10.1128/aem.01043-13)) on these nasal swabs, then processed the sequencing data through [dada2](https://benjjneb.github.io/dada2/). The output we got from dada2 was:

1. An [ASV (amplicon sequencing variant) count table](https://benjjneb.github.io/dada2/) and   
2. Its associated taxonomy table.   

Next, let's load in the example ASV count table. (N.B.: The data doesn't necessarily strictly have to be an ASV table. Any sort of data in tabular format, e.g., OTU table, similar to the example **should** work.)

```{r asvexample}
asvtable <- makeExampleSeqtab()

asvtable
```

Finally, let's load in the example taxonomy file. This was generated during the dada2 workflow, and must match the ASV table.

```{r taxaexample}

taxa <- makeExampleTaxa()

taxa
```

BTW, the function makeExample() also exists for your convenience and will call upon any of the above, like so:

```{r allexample}

metadata <- makeExample("meta")
asvtable <- makeExample("asv")
taxa     <- makeExample("taxa")

```

You can be even more lazy and make all three example tibbles (metadata, ASV table, and its taxonomy table) at once:

```{r allexample2}

all <- makeExample() # OR all <- makeExample("all")

str(all, 1)

# Object "all" is a list, so, you can access (and assign, if you want) each tibble with the $ operator:

metadata <- all$metadata
asvtable <- all$asvtable
taxa     <- all$taxa

```

### Sanity check the data

Now that we've loaded in our main objects (metadata, ASV counts, ASV taxonomy) we will do some basic sanity checks on these three objects.    

First, let's check the metadata object: 

```{r metacheck}

checkMeta(metadata)


```

You'll see there's a warning that NAs were detected in the metadata table. This is not bad or wrong, and it's OK to have NAs! The warning is there to check with *you* that *you* were expecting to see some missing data. If you weren't, check that your metadata object was loaded in correctly.    

In this case, there are NAs in a few spots:

1. Demographic data is "missing" for the lab positive and negative controls, as that kind of information is not relevant for those samples.    
2. Location is missing for the participant that sample "HC1" was taken from. Let's say that the participant declined to share their geographic location, so it is indeed truly missing. With real life data, it's pretty normal to have some information missing. People much smarter than I am have [written](https://hbiostat.org/rmsc/missing.html) [extensively](https://link.springer.com/chapter/10.1007/978-1-4757-3462-1_3) [about](https://pubmed.ncbi.nlm.nih.gov/20338724/) [missing](https://www.appliedmissingdata.com) [data](https://pubmed.ncbi.nlm.nih.gov/17401454/), and how to deal with it, so that won't be discussed much more here.       

This is all fine and there are no glaring red flags with our metadata object.

___

OK, now let's check on ASV and taxonomy tables:

```{r asvcheck}

checkASV(asvtable, taxa, metadata)


```

## Placeholder

This is for me to remember to add to this document:    

Add "bad" examples.       
identifyNegs and assessNegs.       
filtering.     
calcBetaDiv.       


___


# Acknowledgements

As mentioned above, I could not have done any of this without the benefit of what the work of others. These tools below were especially important:

* [mothur](https://mothur.org)   
* [qiime](https://qiime2.org)   
* [mgsat](https://github.com/andreyto/mgsat)   
* maaslin[2](https://huttenhower.sph.harvard.edu/maaslin/)/[3](https://huttenhower.sph.harvard.edu/maaslin3/)   
* [vegan](https://cran.r-project.org/web/packages/vegan/index.html)
* [tidyverse](https://tidyverse.tidyverse.org), especially [dplyr](https://dplyr.tidyverse.org) and [ggplot2](https://ggplot2.tidyverse.org). In addition to these packages, Hadley Wickham and team have written several extremely helpful [guides and tutorials](https://hadley.nz) on data science.
* [dada2 and decontam](https://callahanlab.cvm.ncsu.edu/software/)
* [Suite from Dr. Frank Harrell](https://hbiostat.org), especially [rms](https://cran.r-project.org/web/packages/rms/index.html) and [Hmisc](https://cran.r-project.org/web/packages/Hmisc/index.html).
* Colleague and physician-scientist [Dr. Christian Rosas-Salazar](https://pediatrics.vumc.org/person/christian-rosas-salazar-md-mph) is talented at many things, but has an especial knack for creating figures that are both beautiful and informative.   

More acknowledgements and more details to be added later
