---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# micromeg

<!-- badges: start -->
<!-- badges: end -->

The goal of micromeg is to document and share convenience functions I personally use frequently for microbiome data analysis and processing. This package does not create anything new and is built on the great work of others. Many of the functions it provides already exist in other packages. For example, [phyloseq](https://bioconductor.org/packages/release/bioc/html/phyloseq.html) provides many tools for analysis of microbiome data; it is very well-documented, heavily used (I use it myself!) and so may be better for your purposes.    
    
Other R packages heavily used here include:    
1. [tidyverse](https://tidyverse.tidyverse.org) ecosystem   
2. [vegan](https://cran.r-project.org/web/packages/vegan/index.html)   
3. [dada2](https://benjjneb.github.io/dada2/)   
4. [maaslin3](https://huttenhower.sph.harvard.edu/maaslin3/)        




## Installation

You can install the development version of micromeg from [GitHub](https://github.com/extrasmallwinnie/micromeg/) with:

``` r
# install.packages("pak")
pak::pak("extrasmallwinnie/micromeg")
```

## Example

First, load in some example data to get an idea for the format of what's expected.     
     
I've made up an example study where nasal swabs were taken from people who were either "healthy" or "sick" at the time of sampling. We've collected their health status, their age at collection, sex, and location (rural or urban). Let's load it into our session as an object called "metadata":

```{r examplemeta}
library(micromeg)

metadata <- makeExampleMeta()

metadata

```

This metadata object is a tibble (because I personally like the [tidyverse grammar](https://www.tmwr.org/tidyverse)), but data frames should also be fine to use.     
    
The "SampleID" field hasn't been discussed yet, but it's exactly what it sounds like, and it must be unique, and it must match your sequencing data.    
    
In this made up example, we did 16S sequencing targeting the V4 region (following [this protocol](https://journals.asm.org/doi/10.1128/aem.01043-13)) on these nasal swabs, then processed the specimens through the [dada2](https://benjjneb.github.io/dada2/) pipeline. The output we get from dada2 is:

1. An [ASV (amplicon sequencing variant) table](https://benjjneb.github.io/dada2/).   
2. Its associated taxonomy table.   

Next, let's load in the example [ASV table](https://benjjneb.github.io/dada2/). (N.B.: Any sort of data in tabular format similar to the sample **should** work, but no guarantees.)

```{r asvexample}
asvtable <- makeExampleSeqtab()

asvtable
```

You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date. `devtools::build_readme()` is handy for this.

You can also embed plots, for example:

```{r pressure, echo = FALSE}

```

In that case, don't forget to commit and push the resulting figure files, so they display on GitHub and CRAN.
